From a3dd7d2607601f3adba6962bf3d9e49f2b4edfb0 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 26 Aug 2014 14:12:04 -0400
Subject: [ethernet] igb: fix stats for i210 rx_fifo_errors

Message-id: <1409062359-13063-47-git-send-email-sassmann@redhat.com>
Patchwork-id: 89188
O-Subject: [RHEL7.1 PATCH 46/81] igb: fix stats for i210 rx_fifo_errors
Bugzilla: 1091121
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Todd Fujinaka <todd.fujinaka@intel.com>

RQDPC on i210/i211 is R/W not ReadClear. Clear after reading.

Signed-off-by: Todd Fujinaka <todd.fujinaka@intel.com>
Tested-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit e66c083aab32842f225bae2a2c30744bf96abaec)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/igb/igb_main.c b/drivers/net/ethernet/intel/igb/igb_main.c
index 5d4dbe5..e149d37 100644
--- a/drivers/net/ethernet/intel/igb/igb_main.c
+++ b/drivers/net/ethernet/intel/igb/igb_main.c
@@ -5193,8 +5193,10 @@ void igb_update_stats(struct igb_adapter *adapter,
 
  rcu_read_lock();
  for (i = 0; i < adapter->num_rx_queues; i++) {
-  u32 rqdpc = rd32(E1000_RQDPC(i));
   struct igb_ring *ring = adapter->rx_ring[i];
+  u32 rqdpc = rd32(E1000_RQDPC(i));
+  if (hw->mac.type >= e1000_i210)
+   wr32(E1000_RQDPC(i), 0);
 
   if (rqdpc) {
    ring->rx_stats.drops += rqdpc;
-- 
1.7.1